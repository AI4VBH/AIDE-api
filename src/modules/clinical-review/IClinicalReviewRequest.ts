export interface IClinicalReviewRequest {
  pageNumber: number;
  pageSize: number;
  patientId: string;
  patientName: string;
  applicationName: string;
  roles: string[];
}
