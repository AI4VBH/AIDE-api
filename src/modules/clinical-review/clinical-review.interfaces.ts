import { IMonaiPayloadPatient } from 'modules/admin/payloads/payload.interface';
import { IPagedResponse } from 'shared/helper/paging/paging.interface';

export type PagedClinicalReviews = IPagedResponse<ClinicalReview>;

export interface PagedClinicalReviewDto {
  totalPages: number;
  totalRecords: number;
  data: ClinicalReview[];
}

export interface ClinicalReview {
  task_id: string;
  reviewed_task_id: string;
  execution_id: string;
  reviewed_execution_id: string;
  correlation_id: string;
  workflow_name: string;
  patient_metadata: IMonaiPayloadPatient;
  files: [{ [name: string]: string }];
}

export interface ClinicalReviewAcknowledge {
  acceptance: boolean;
  task_id: string;
  reason: string;
  message: string;
  execution_id: string;
  roles: string[];
  userId: string;
}

export interface ClinicalReviewTaskDetails {
  execution_id: string;
  study: ClinicalReviewTaskSeries[];
}

export interface ClinicalReviewTaskSeries {
  series_id: string;
  modality: string;
  files: string[];
}
