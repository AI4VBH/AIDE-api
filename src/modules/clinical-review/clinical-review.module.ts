import { Module } from '@nestjs/common';
import { ElasticModule } from '../../shared/elastic/elastic.module';
import { ExecutionsController } from './executions/executions.controller';
import { ExecutionsService } from './executions/executions.service';

@Module({
  imports: [ElasticModule],
  controllers: [ExecutionsController],
  providers: [ExecutionsService],
})
export class ClinicalReviewModule {}
