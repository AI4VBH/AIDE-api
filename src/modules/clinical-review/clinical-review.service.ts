/*
 * Copyright 2022 Crown Copyright
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { firstValueFrom } from 'rxjs';
import { ConfigService } from '@nestjs/config';
import {
  ClinicalReviewAcknowledge,
  PagedClinicalReviews,
} from './clinical-review.interfaces';

@Injectable()
export class ClinicalReviewService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  @Inject(ConfigService)
  private readonly configService: ConfigService;

  async getClinicalReviews(
    pageNumber: number,
    pageSize: number,
    roles: string[],
  ) {
    const params = new URLSearchParams({
      pageSize: `${pageSize || ''}`,
      pageNumber: `${pageNumber || ''}`,
      roles: `${roles.join(',') || ''}`,
    });

    const baseURL = this.configService.get<string>(
      'CLINICAL_REVIEW_SERVICE_HOST',
    );

    const getClinicalReviews = await firstValueFrom(
      this.httpService.get<PagedClinicalReviews>(`/clinical-review?${params}`, {
        baseURL,
      }),
    );

    return getClinicalReviews.data;
  }

  async acknowledgeClinicalReview(
    clinicalReview: ClinicalReviewAcknowledge,
    roles: string[],
    userId: string,
    clinicalReviewId: string,
  ) {
    clinicalReview.roles = roles;
    clinicalReview.userId = userId;

    const baseURL = this.configService.get<string>(
      'CLINICAL_REVIEW_SERVICE_HOST',
    );

    const acknowledgeClinicalReview = await firstValueFrom(
      this.httpService.put(
        `/clinical-review/${clinicalReviewId}`,
        clinicalReview,
        {
          baseURL,
        },
      ),
    );

    return acknowledgeClinicalReview.data;
  }
}
