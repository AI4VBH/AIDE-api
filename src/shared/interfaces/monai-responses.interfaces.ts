export interface IMonaiProblemResponse {
  type: string;
  title: string;
  status: number;
  detail: string;
  instance: string;
  traceId: string;
}
