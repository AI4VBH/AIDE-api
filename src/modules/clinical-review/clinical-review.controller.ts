import {
  BadRequestException,
  Body,
  Controller,
  Get,
  HttpCode,
  HttpStatus,
  Inject,
  Param,
  ParseIntPipe,
  Put,
  Query,
  UseFilters,
} from '@nestjs/common';
import { Roles, UserId } from 'shared/decorators/custom-decorators';
import ExternalServerExceptionFilter from 'shared/http/external-server-exception.filter';
import {
  ClinicalReviewAcknowledge,
  ClinicalReviewTaskDetails,
} from './clinical-review.interfaces';
import { ClinicalReviewService } from './clinical-review.service';

@Controller('clinical-review')
@UseFilters(ExternalServerExceptionFilter)
export class ClinicalReviewController {
  @Inject(ClinicalReviewService)
  private readonly service: ClinicalReviewService;

  @Get()
  getClinicalReviews(
    @Roles() roles,
    @Query('pageNumber', ParseIntPipe) pageNumber = 1,
    @Query('pageSize', ParseIntPipe) pageSize = 10,
  ) {
    if (pageNumber <= 0 || pageSize <= 0) {
      throw new BadRequestException('pageNumber or pageSize is invalid');
    }

    if (!roles) {
      throw new BadRequestException('roles are required');
    }

    return this.service.getClinicalReviews(pageNumber, pageSize, roles);
  }

  @Put(':clinicalReviewId')
  @HttpCode(HttpStatus.NO_CONTENT)
  acknowledgeClinicalReview(
    @Param('clinicalReviewId') clinicalReviewId: string,
    @Body() clinicalReview: ClinicalReviewAcknowledge,
    @Roles() roles,
    @UserId() userId,
  ) {
    if (!roles || !userId) {
      throw new BadRequestException('roles and userId are required');
    }

    return this.service.acknowledgeClinicalReview(
      clinicalReview,
      roles,
      userId,
      clinicalReviewId,
    );
  }

  @Get('dicom')
  GetDicomFile(@Query('key') key: string) {
    return this.service.getDicomFile(key);
  }

  @Get(':taskExecutionId')
  GetClinicalReviewTaskDetails(
    @Roles() roles,
    @Param('taskExecutionId') taskExecutionId: string,
  ): Promise<ClinicalReviewTaskDetails> {
    if (!roles) {
      throw new BadRequestException('roles are required');
    }

    return this.service.getClinicalReviewTaskDetails(roles, taskExecutionId);
  }
}
