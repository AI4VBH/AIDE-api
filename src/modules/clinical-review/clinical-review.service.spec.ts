import { createMock, DeepMocked } from '@golevelup/ts-jest';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { AxiosInstance } from 'axios';
import { makeObservableForTest } from '../../../test/utilities/test-make-observable';
import * as clinicalReviewInterfaces from './clinical-review.interfaces';
import { ClinicalReviewService } from './clinical-review.service';

describe('WorkflowsService', () => {
  let axios: DeepMocked<AxiosInstance>;
  let httpService: DeepMocked<HttpService>;
  let configService: DeepMocked<ConfigService>;

  let service: ClinicalReviewService;

  beforeEach(async () => {
    axios = createMock<AxiosInstance>();
    httpService = createMock<HttpService>({
      axiosRef: axios,
    });
    configService = createMock<ConfigService>();

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        ClinicalReviewService,
        {
          provide: HttpService,
          useValue: httpService,
        },
        {
          provide: ConfigService,
          useValue: configService,
        },
      ],
    }).compile();

    service = module.get<ClinicalReviewService>(ClinicalReviewService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getClinicalReviews', () => {
    it('returns expected result', async () => {
      const monaiResult: clinicalReviewInterfaces.PagedClinicalReviews = {
        totalPages: 1,
        totalRecords: 0,
        pageNumber: 1,
        pageSize: 10,
        firstPage: '/page-1',
        lastPage: '/page-1',
        nextPage: null,
        previousPage: null,
        data: [],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: monaiResult,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const result = await service.getClinicalReviews(1, 10, ['admin']);

      expect(result).toMatchSnapshot();
    });
  });

  describe('acknowledgeClinicalReview', () => {
    it('returns expected result', async () => {
      const review: clinicalReviewInterfaces.ClinicalReviewAcknowledge = {
        acceptance: true,
        task_id: 'taskid',
        reason: 'all good',
        message: 'message',
        execution_id: 'guid',
        roles: [],
        userId: '',
      };

      const clinicalReviewId = 'clinical-review-id';

      axios.put.mockResolvedValue({
        status: 201,
      });

      httpService.put.mockReturnValue(makeObservableForTest(axios.put));

      const result = await service.acknowledgeClinicalReview(
        review,
        ['admin'],
        'userid',
        clinicalReviewId,
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('getClinicalReviewTaskDetails', () => {
    it('returns expected result', async () => {
      const result: clinicalReviewInterfaces.ClinicalReviewTaskDetails = {
        execution_id: '12345',
        study: [
          {
            series_id: 'series-id',
            modality: 'CT',
            files: ['object-keys-for-minio'],
          },
          {
            series_id: 'another-series-id',
            modality: 'DOC',
            files: [
              'other-object-keys-for-minio',
              'another-object-keys-for-minio',
            ],
          },
        ],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: result,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const clinicalTaskReviewId = 'id-123';
      const roles = ['realm:clinician'];
      const taskDetailsReturned = await service.getClinicalReviewTaskDetails(
        roles,
        clinicalTaskReviewId,
      );

      expect(taskDetailsReturned).toMatchSnapshot();
    });
  });
});
