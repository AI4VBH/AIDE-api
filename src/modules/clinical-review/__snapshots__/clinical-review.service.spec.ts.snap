// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowsService acknowledgeClinicalReview returns expected result 1`] = `undefined`;

exports[`WorkflowsService getClinicalReviewTaskDetails returns expected result 1`] = `
{
  "execution_id": "12345",
  "study": [
    {
      "files": [
        "object-keys-for-minio",
      ],
      "modality": "CT",
      "series_id": "series-id",
    },
    {
      "files": [
        "other-object-keys-for-minio",
        "another-object-keys-for-minio",
      ],
      "modality": "DOC",
      "series_id": "another-series-id",
    },
  ],
}
`;

exports[`WorkflowsService getClinicalReviews returns expected result 1`] = `
{
  "data": [],
  "firstPage": "/page-1",
  "lastPage": "/page-1",
  "nextPage": null,
  "pageNumber": 1,
  "pageSize": 10,
  "previousPage": null,
  "totalPages": 1,
  "totalRecords": 0,
}
`;
